src/hashfunctions.c

```diff
@@ -0,0 +1,273 @@
+                     (Int)TNAM_OBJ(trans), 0L);
+    }
+
+    return HashValueToObjInt(HashFuncForTrans(trans));
+}
+
+Obj DATA_HASH_FUNC_FOR_STRING(Obj self, Obj string)
+{
+    /* check the argument                                                  */
+    if (!IS_STRING(string)) {
+        ErrorMayQuit("DATA_HASH_FUNC_FOR_STRING: <string> must be a "
+                     "string (not a %s)",
+                     (Int)TNAM_OBJ(string), 0L);
+    }
+
+    if(!IS_STRING_REP(string))
+    {
+        string = CopyToStringRep(string);
+    }
+
+
+    UInt len = GET_LEN_STRING(string);
+    UInt1* ptr = CHARS_STRING(string);
+
+    UInt hashval = HASHKEY_MEM_NC(ptr, 2782, len);
+    return HashValueToObjInt(hashval);
+}
+
+Int DataHashFuncForInt(Obj i)
+{
+    GAP_ASSERT(TNUM_OBJ(i) == T_INTPOS || TNUM_OBJ(i) == T_INTNEG);
+    if (TNUM_OBJ(i) == T_INTPOS)
+    {
+        return HASHKEY_WHOLE_BAG_NC(i, 293479);
+    }
+    else
+    {
+        return HASHKEY_WHOLE_BAG_NC(i, 193492);
+    }
+}
+
+Obj DATA_HASH_FUNC_FOR_INT(Obj self, Obj i)
+{
+    /* check the argument                                                  */
+    if (TNUM_OBJ(i) != T_INT && TNUM_OBJ(i) != T_INTPOS &&
+        TNUM_OBJ(i) != T_INTNEG) {
+        ErrorMayQuit(
+            "DATA_HASH_FUNC_FOR_INT: <i> must be an integer (not a %s)",
+            (Int)TNAM_OBJ(i), 0L);
+    }
+
+    if (IS_INTOBJ(i))
+        return i;
+    else
+        return HashValueToObjInt(DataHashFuncForInt(i));
+}
+
+Int BasicRecursiveHash(Obj obj);
+
+Int BasicRecursiveHashForList(Obj obj)
+{
+    GAP_ASSERT(IS_LIST(obj));
+    // This is just a randomly chosen number which fits in a 32-bit integer
+    UInt current_hash = 2195952830;
+    Int  len = LEN_LIST(obj);
+    for (Int pos = 1; pos <= len; ++pos) {
+        Obj val = ELM0_LIST(obj, pos);
+        if (val == 0) {
+            current_hash = HashCombine2(current_hash, ~(Int)0);
+        }
+        else {
+            current_hash =
+                HashCombine2(current_hash, BasicRecursiveHash(val));
+        }
+    }
+    return current_hash;
+}
+
+Int BasicRecursiveHashForPRec(Obj obj)
+{
+    GAP_ASSERT(IS_PREC_REP(obj));
+
+    UInt current_hash = 1928498392;
+
+    /* ensure record is sorted */
+    SortPRecRNam(obj, 0);
+
+    /* hash componentwise                                               */
+    for (Int i = 1; i <= LEN_PREC(obj); i++) {
+        UInt recname = GET_RNAM_PREC(obj, i);
+        UInt rechash = BasicRecursiveHash(GET_ELM_PREC(obj, i));
+        current_hash = HashCombine3(current_hash, recname, rechash);
+    }
+
+    return current_hash;
+}
+
+Int BasicPrimitiveHash(Obj obj)
+{
+    if (IS_INTOBJ(obj)) {
+        return (Int)obj;
+    }
+    else if (TNUM_OBJ(obj) == T_CHAR) {
+        return *(UChar *)ADDR_OBJ(obj);
+    }
+    else if (TNUM_OBJ(obj) == T_BOOL) {
+        if (obj == True)
+            return 1;
+        else if (obj == False)
+            return 2;
```

Comment:
So... 0 and True hash to the same value. And small integers 0-255 hash to the same values as chars. Do we really want that?